### DISCUSSION TITLE
Language Specification: statics and inheritance

### DISCUSSION THREAD

**Discussion by errael** [vim/vim#13004] [General]
(Closed as of patch 9.0.1906 https://github.com/vim/vim/commit/92d9ee5f4ca0d2de04c39afbafc7609da43fb2e9)

Originally was going to name this discussion "interface statics". The idea was to explore what a static variable declared in an interface means. #12987 discusses members in an interface and what they mean; one way to thing about them is as properties.
When setting up some example I ran into something that may be a bug

**Simple `extends` with `static`**
```
class A
    public static pu_class_var: number
endclass

class B extends A
endclass

class C extends A
endclass
```
This example compiles (doesn't do much). How many statics are there? I would have guessed 1; there are 3. Is this a bug? In the next message there is a test program that demonstrates this.

The initial idea of this discussion is to study and clarify what a static in an interface means, but sidetracked by previous example.
**Simple `implements` with `static`**
```
vim9script

interface A
    public static pu_class_var: number
endinterface

class B implements A
endclass

class C implements A
endclass
```
It doesn't compile because both B and C need to implement `pu_class_var`. An argument can be made that there is only one static and that the classes that implement `A` do not need to implement `pu_class_var`; _not the generally accepted interpretation_.

I did not see anything about either of these situations in the spec, but that doesn't mean it's not there

**Comment by errael** [comment]
@yegappan @chrisbra @dkearns looking for _Language Lawyers_

Example **Simple extends with static**. Using getters and setters to avoid known bugs. The output of the script is
```
A: 1, B: 2, C: 3
```
Clearly showing that there are 3 separate static variables.

```
vim9script

class A
    public static pu_class_var: number
    def WA(arg: number)
        pu_class_var = arg
    enddef
    def RA(): number
        return pu_class_var
    enddef
endclass

class B extends A
    #public static pu_class_var: number
    def WB(arg: number)
        pu_class_var = arg
    enddef
    def RB(): number
        return pu_class_var
    enddef
endclass

class C extends A
    #public static pu_class_var: number
    def WC(arg: number)
        pu_class_var = arg
    enddef
    def RC(): number
        return pu_class_var
    enddef
endclass

var oa = A.new()
var ob = B.new()
var oc = C.new()

def Vals()
    var a = printf("A: %d, B: %d, C: %d", oa.RA(), ob.RB(), oc.RC())
    echo a
enddef

def F()
    oa.WA(1)
    ob.WB(2)
    oc.WC(3)
    Vals()
enddef
F()
```

**Reply by yegappan** [reply]
The documentation states the following:

```
Object members from the base class are all taken over by the child class.  It
is not possible to override them (unlike some other languages).
```

When a class extends the parent, all the members of the parent are taken
over by the child class.  So a copy needs to be made as the class can be
extended by multiple children.

**Reply by errael** [reply]
I wish there was some examples and/or discussion of extends that had statics. `vim9script` typically follows common conventions and where it doesn't the spec often says why it doesn't. (unless I'm just saying that to argue my point). This just seems so wrong, it will take some time for me to digest this. In Java for example, in a class a static variable is inherited/shared, not copied; however they may be "hidden" by declaring something with the same name. Of course, `vim9script` doesn't want complex rules like hiding, and takes a simpler course; I wonder if that simpler course is to **not allow shadowing** instead of **requiring shadowing** which is how it works now. What's even more confusing for me is that `vim9script` typically does not want there to be shadowing.

Have you seen any comments in the code about inheriting statics? That would be good for me to look at; maybe calm me down a bit.

**Reply by dkearns** [reply]
I charge by the word and photocopied page for my ramblings so I'll try and keep this short.

I'd also expect this to return "3, 3, 3" or an error for an undeclared pu_class_var in the subclasses.

I am certainly surprised to see that a shadowing static field is declared.  Generally the static field is inherited or it's not, this seems a bit like black magic.  It would probably be hard to discern what was happening without poring through the documentation.

In terms of some 'mainstream' languages, and off the top of my head, I think Java, C#, PHP, Ruby, Python, and JavaScript inherit while Dart does not.  Python and JavaScript arguably require a bit of syntactic hoop jumping if you want to set the super-class static field rather than create a shadowing one in subclasses and Dart requires you to explicitly reference the super class by name but they're all able to do this.

It looks like one can't currently assign to the super-class variable from a subclass?  We can't simulate this with getters and setters if that's case but I'd hope that was a bug.

I suppose script-local variables can be used to achieve much the same thing.

Relax, it's only software. :)

**Reply by errael** [reply]
> surprised to see that a shadowing static field is declared
> Generally the static field is inherited or it's not, this seems a bit like black magic.
>  It would probably be hard to discern what was happening 

Seems to violate the principal of _least surprise_.

> [... inherits, but Dart does not ... no way to reference superclass value unlike any other language]

You **are** a _Language Lawyer_. It's interesting that `Dart` doesn't inherit, though it does allow super. I seem to recall that Bram took some inspiration from `Dart`.

Interestingly I suspect it would be _easy_ to change. Just copy the `typeval_T` pointer, instead of doing `copy_tv`; the classes would literally share the value; wouldn't need new code for searching up the hierarcy; shadowing would be forbidden. But I'm just guessing, haven't looked at code.

> suppose script-local variables can be used to achieve much the same thing

Except for access control.

Anyone have use-cases for either way of doing this?

**Reply by dkearns** [reply]
> Seems to violate the principal of least surprise.

I had quick look at some others I wasn't sure about and it seems D, FreeBasic, Delphi, and C++ inherit as well.

> It's interesting that Dart doesn't inherit, though it does allow super. I seem to recall that Bram took some inspiration from Dart.

I didn't really read it but there's some discussion [here](https://groups.google.com/a/dartlang.org/g/misc/c/dudRfe0YSyo) - Ahé is the designer.   Other languages like Kotlin, while a bit different, also disallow it.

The theory goes that there's a long history of confusion in the Java world. :)

**Reply by errael** [reply]
> The theory goes that there's a long history of confusion in the Java world. :)

It's more than just theory, AFAICT.

So static inheritance is all over the map. Unless @yegappan has a changing  opinion, **it stays like it is**. I can't think of a technical reason, or use cases, favoring one or the other; it's more about the "least surprise".

> a bit like black magic

Perhaps every language needs just a bit.

**Reply by dkearns** [reply]
@errael, just to be clear Vim9 isn't doing either.  IUUC, it's creating a shadowing static variable initialised to the initial value of the parent class. 

FWIW, I'm deliberately not looking at the implementation to see if I can understand these choices from the casual user's perspective.

@yegappan, just to flog a dead horse can you please confirm that the following is correct and possibly explain the motivation for eschewing popular language norms with these semantics?  Is there some common use for this behaviour?  If so, this is going to need some explanation under `:help static`.

```vim
vim9script

class A
  public static foo = 42
endclass

A.foo = 123

class B extends A
endclass

echomsg B.foo
A.foo = 456
echomsg B.foo
```

Output:
```
42
42
```

If it doesn't meet the first of goal of Vim9 and improve performance how does it meet the secondary goal:

> to avoid Vim-specific constructs and get closer to commonly used programming languages, such as JavaScript, TypeScript and Java.

**Reply by errael** [reply]
> just to be clear Vim9 isn't doing either ... 

Thanks for clarifying. I re-opened this discussion.

**Reply by yegappan** [reply]
In the current implementation of Vim9 classes (as you described above), when a class is extended, the class member variables are copied
from the parent class to the child class along with their initialization expression (if any).  So changing the value of the parent class
member variable doesn't have any impact on the child class member variable.

In the example you quoted above:

```
vim9script
class A
  public static foo = 42
endclass

A.foo = 123

class B extends A
endclass
```

Class B gets a copy of `foo` and is initialized with the specified initialization expression `= 42` in the parent class.
The current value of `foo` (which is 123) is not used.  So the class member `B.foo` is initialized to 42.

```
echomsg B.foo
```

As the current value of `B.foo` is 42, this results in 42.

```
A.foo = 456
```

Changing the value of the class member `A.foo` doesn't change `B.foo`.

```
echomsg B.foo
```

Here `B.foo` is still 42, so this results in 42.

The documentation currently has the following:

```
Object members from the base class are all taken over by the child class.  It
is not possible to override them (unlike some other languages).
```

If this behavior needs to be changed, then we need to discuss what the new behavior should be.

**Reply by dkearns** [reply]
FWIW, I read that help entry as only referring to (instance) member variables and unrelated to class members.  E.g., so that the following is not allowed, as it commonly is by other languages.

```vim
class A
  this.foo: number
endclass
class B extends A
  this.foo: string
endclass
```

Shadowing static field declarations in subclasses should definitely be allowed.  If a shadowed declaration does not exist in the subclass it's common to allow for a superclass static members to also be referenced by the subclass name.

So the example should output:
```
123
456
```

I think the ability to access the superclass static members via the subclass name is a bit pointless.  Simplifying this so that static members must be accessed by the name of the class that declares them would be an improvement and seems to be the approach taken by some of the more modern languages.  Then the only exception would be that from within the declaring class they can be referenced by the bare identifiers.

**Reply by errael** [reply]
> Shadowing static field declarations in subclasses should definitely be allowed.

Why?

**Reply by dkearns** [reply]
Are you asking why this should not be allowed?

```vim
class A
  static foo: int
endclass
class B extends A
  static foo: int
endclass
```

The common language around static members is confusing because the feature was included, in popular languages, at a time when the fad was "objects and inheritance"...assuming I have my timeline roughly right.

**Reply by errael** [reply]
> Are you asking why this should not be allowed?

I'm asking why it should be allowed. (not not)

Currently there's a mysterious static inheritance scheme. I believe the subclasses should share the base class static. What's the advantage or use-case in allowing shadowing here?

**Reply by dkearns** [reply]
Sorry I mistyped...

Well, when you don't want to share you can shadow. :)

If we're in agreement that we should do what most relevant languages do then that is represented by the following.  On the other hand some like Dart, for example, wouldn't allow `B` to access `foo` - they don't care to share.

TypeScript

```tsx
class A {
	static foo: number = 11
	static bar: number = 22
}
class B extends A {
	static bar: number = 33
}
console.log(A.foo)
console.log(A.bar)
console.log(B.bar)
```
```
[LOG]: 11 
[LOG]: 22 
[LOG]: 33 
```

[TS Playground Link](https://www.typescriptlang.org/play?&q=293#code/MYGwhgzhAECC0G8BQBICAXM6CWxoDMB7QgLmgDsBXAWwCMBTAJ2gF5oBGd1DLXaWsIzJU6TVtABMEpAF8koSDABC0egA909cgBMY8ZGkw48AoRRoNmbAMzXZ8wuQiEQ9AHQhCAcwAUsN0SEAJQOTi7unr7+piHAjs6uHt4+Sm4xQA)

**Reply by errael** [reply]
> Well, when you don't want to share you can shadow. :)

(that's compelling, but I'm a sucker for a good turn of phrase)

You're saying you can get the current semantics by explicitly declaring the shadow.

> If we're in agreement that we should do what most relevant languages do

I think we'd agree that we shouldn't do something semantically different from what any other language does (like the current static member inheritance). But "should do what most relevant languages do", I'm not sure I'd go that far. Shadowing variables is clearly one place where vim diverges from many languages; part of it's KISS approach I suppose.

Since allowing static shadowing by declaration at a later date is compatible, I'd defer a decision until there's some experience/demand/use-case showing that it's needed.

**Reply by yegappan** [reply]
I see that in typescript, a static variable in an extended class is cloned when the variable is modified using the subclass:

```
class A {
	static bar: number = 22
}
class B extends A {
}
console.log(A.bar)
console.log(B.bar)
A.bar = 33
console.log(A.bar)
console.log(B.bar)
B.bar = 44
console.log(A.bar)
console.log(B.bar)
A.bar = 55
console.log(A.bar)
console.log(B.bar)
```

The above produces the following output:
```
22
22
33
33
33
44
55
44
```

As you can see from the above output, after the `B.bar = 44` statement, the `bar` variable in `B` is no longer referring to the one in `A`.
This behavior is confusing (because before the modification, `B.bar` was referring to the value of `A.bar`).

**Reply by yegappan** [reply]
For comparison, in Java, a similar program produces the below output.

MyClass.java:
```
public class MyClass {
    public static void main(String args[]) {
        System.out.println("A.bar = " + A.bar);
        System.out.println("B.bar = " + B.bar);
        A.bar = 33;
        System.out.println("A.bar = " + A.bar);
        System.out.println("B.bar = " + B.bar);
        B.bar = 44;
        System.out.println("A.bar = " + A.bar);
        System.out.println("B.bar = " + B.bar);
        A.bar = 55;
        System.out.println("A.bar = " + A.bar);
        System.out.println("B.bar = " + B.bar);
    }
}
```

A.java:
```
class A
{
    static int bar = 22;
}
```

B.java:
```
class B extends A
{
}
```

Output:
```
A.bar = 22
B.bar = 22
A.bar = 33
B.bar = 33
A.bar = 44
B.bar = 44
A.bar = 55
B.bar = 55
```

In this case, both `A.bar` and `B.bar` refer to the same variable and produce the same value.

**Reply by yegappan** [reply]
In Dart, static members are not inherited and cannot be referenced using a subclass: https://dart.dev/guides/language/specifications/DartLangSpec-v2.10.pdf#subsection.10.7

**Reply by yegappan** [reply]
Interestingly, Python3 behaves similar to Typescript:

```
class A:
    bar = 22

class B(A):
    pass

print(f"A.bar = {A.bar}")
print(f"B.bar = {B.bar}")
A.bar = 33
print(f"A.bar = {A.bar}")
print(f"B.bar = {B.bar}")
B.bar = 44
print(f"A.bar = {A.bar}")
print(f"B.bar = {B.bar}")
A.bar = 55
print(f"A.bar = {A.bar}")
print(f"B.bar = {B.bar}")
```

The above program produces the following output:
```
A.bar = 22
B.bar = 22
A.bar = 33
B.bar = 33
A.bar = 33
B.bar = 44
A.bar = 55
B.bar = 44
```

**Reply by dkearns** [reply]
In TypeScript (and similarly in Python) it's not really cloning in that the same-named static variable of A is given no consideration.  It's just creating a new static variable on B - sometimes referred to as shadowing to distinguish it from overriding.  It is the same as declaring it within the class definition with `static bar = 44`.

`B.baz = 99` would also create a new `baz` static variable on B that obviously has no relationship with A.

That's the 10,000ft view, there's other subtleties here but they don't apply to Vim9 script.

Java is not scripty enough for this and requires the declaration.  I think we're generally much closer to it in behaviour.

I don't think static members are particularly useful in a language with modules and closures but I can't squint hard enough to pretend it's not there in the spec and I suppose there are some name-spacing benefits.

So. I'd vote for Dart's approach (assuming voting was free).  If you want to use the thing just use the thing.  E.g., if you want to reference `A.bar` from the subclass just use `A.bar`.

> In Dart, static members are not inherited and cannot be referenced using a subclass: https://dart.dev/guides/language/specifications/DartLangSpec-v2.10.pdf#subsection.10.7

I agree with the slightly snarky commentary here - seems like a fun spec.

**Reply by errael** [reply]
`vim9script` has a strict no shadowing rule. I don't see any reason to change that. What's the advantage?

The bizarre, magical, and surreptitious behavior of current static inheritance should be fixed; the simplest seems shared. What are other specific proposals for `vim9script`. https://github.com/vim/vim/discussions/13004#discussioncomment-6929014 is the shared one. If a proposal can't be stated simply, then I'm not sure it fits with `vim9script`.

**Reply by dkearns** [reply]
How can sharing be the simplest?  It's right there in the name that it requires an extra mechanism, sharing. :)

Sharing without 'shadowing' would be unique, to my knowledge, and for good reason.  Why should a static member in a superclass prevent a same-named static member in a subclass.  There's no significant relationship between the two and they're not in the same scope.

Sharing is useless and _creates_ the 'shadowing' complication. Without sharing there can be no 'shadowing' because the static member is exclusively name-spaced to the class - all access is qualified by the class name.

To reiterate the simplest static member option is to only allow them to be accessed by class name or bare identifier from within the defining class(*). 

```vim
class A
  static foo = 42
endclass

# now the only way to access foo is as A.foo
```

The simplest solution overall is to nuke static members from orbit.  It's the only way to be sure.

I'm going to start referring to this as the Care Bear feature...


`*` This is inconsistent with other Vim9 syntax rules but that's another hill to die on

**Reply by errael** [reply]
> simplest static member option is to only allow them to be accessed by class name or bare identifier from within the defining class

I'm OK with that. Subclasses can't access superclass static with bare reference (i.e. no sharing), no shadowing. In the hierarchy, first come, first serve.

**Reply by errael** [reply]
Extending a class
```
Object members from the base class are all taken over by the child class.  It
is not possible to override them (unlike some other languages).

Class members are not inherited, outside the class the name must be prefixed.
It is not  possible to override them (unlike some other languages).
```

**Reply by errael** [reply]
Special handling for a static named `lv_426`, exception: `E0000:  Express Elevator To Hell, Goin' Down`

**Comment by yegappan** [comment]
The Vim9 class documentation states that all the items specified in an interface must appear in the class:
```
If a class declares to implement an interface, all the items specified in the
interface must appear in the class, with the same types. *E1348* *E1349*

An interface can declare object members, just like in a class but without any
initializer.
```

So it is good to check at compile time whether all the items in the interface are
implemented or not.

Multiple classes can implement an interface.  So the static members in an interface
must be copied to the class.  It is expected that you will have one static member
for every class that implements the interface.

**Reply by errael** [reply]
```
If a class declares to implement an interface, all the items specified in the
interface must appear in the class, with the same types. *E1348* *E1349*

An interface can declare object members, just like in a class but without any
initializer.
```
I see that. I'm troubled that there are no examples of declaring a static in an interface.
In some languages, a static in an interface _does not_ need to be implemented. I'm not
saying that vim9script should follow these; only that there are examples of where
a static in a interface is shared by all the implementers.

Can you describe a use-case where a static is in an interface and you want each
class to have it's own static?

> Multiple classes can implement an interface. So the static members in an interface
> must be copied to the class

**Comment by errael** [comment]
I'm trying to think of use cases and how I would do things. I guess the bottom line is that _if you extend something, you do not share any of its state_. I haven't run any tests, can statics in super classes be referenced? Like super.some_static_var?

I guess for sharing "statics" the thing to do is set up a class that only declares statics and is never extended and share that.

**Comment by errael** [comment]
**An alternate static inheritance mechanism**

`class members` handling is not specified for `extends`. The following spec addition/clarification goes for sharing.
There are other possibilities. 

```
Object members from the base class are all taken over by the child class.  It
is not possible to override them (unlike some other languages).

Class members (AKA statics) from the base class are shared by the child class.
It is not  possible to override them (unlike some other languages).
```

**Comment by errael** [comment]
I can't help myself, I have to look at implementation issues. Earlier in this discussion, I'd said

>  I suspect it would be easy to change. Just copy the typeval_T pointer, instead of doing copy_tv

In fact there is an array of `tv` not an array of pointer to `tv`. So not quite as simple as I'd hoped.

Considering a `shared statics` specification, I first did
```
$ grep class_members_tv *.c
```
I didn't see that many, so I wondered about introducing into class_T 
```
    typval_T	**p_class_members_tv;	// allocated array ptr to vals
```
Thanks to the refactoring of `ex_class` that @yegappan did, I isolated the primary change to `add_class_members()`. There's a draft PR #13045 (not anywhere close to a full implementation) exploring this. It wastes storage with `class_members_tv ` but not very much. The implementation minimizes structural changes; and regardless of that, might the right choice structurally. If there's interest, I could try fleshing this out.

**Comment by errael** [comment]
This discussion was originally conceived to discuss `interface static` semantics. `class static` inheritance was added and has gotten the bulk of the air time.

Should a static in an interface, if statics in interface are supported, be automatically `const`?

**Reply by dkearns** [reply]
Is this documented somewhere?   I can't find it, so that seems like an opportunity. :)

**Reply by errael** [reply]
> Is this documented somewhere?

There's not much about `class member` in the spec. But, upon reflection, the issue:"if statics in interface are supported" is entwined with "if they are supported, how are they inherited" which is closely related to the `class static` inheritance question.

**Reply by errael** [reply]
> if statics in interface are supported

At this point in time, unless there's a good use case, I'd probably not allow it for vim9.1.

It's easier to add something than take it away.

**Reply by errael** [reply]
There's "read between the lines" evidence that statics in an interface don't need to be supported.

In the spec, see `:he specifies` it says
```
A class can declare its interface, the object members and methods, with a
named interface.  This avoids the need for separately specifying the
interface, which is often done in many languages, especially Java.
```
Note that `class members and methods`, aka `static`, are not mentioned. Is the implication that interfaces don't have statics, or simply that they're not part of the interface? Or maybe there's no implication because statics are not fully specified.

**Comment by errael** [comment]
@yegappan Any comment on https://github.com/vim/vim/discussions/13004#discussioncomment-6929014

Doing a search in vim9class.txt for `class member\|object member` there's lots of `object member` and few `class member`.

**Comment by errael** [comment]
Updated the intro to this discussion #13004.

There two proposals so far for _Simple `extends` with `static`_.

Haven't heard much about _Simple `implements` with `static`_. As is? No preferences? Nuke them? Const?

There's a weird claim that the spec implies that an interface can't have statics.
https://github.com/vim/vim/discussions/13004#discussioncomment-6962590

Note that anything that anything that is compatible with the current spec,
can be added after vim9.1 is released.

**Comment by yegappan** [comment]
I ran some tests using sample Java, Typescript and Dart programs to understand how the class variables are shared between a parent and a child class and how they can be accessed from class/object methods.

[Typescript Sample Program](https://www.typescriptlang.org/play?&q=293#code/MYGwhgzhAECC0G8BQBICAXM6CWxoDMB7QgLmgDsBXAWwCMBTAJ2gF5oBGVDLXaWsRmSp0mrDgAZOaTDjxgA+hAAKjbOXQAKAJSpkKFAHoDwQuQiEQ9AHQhCAcw0AiBctXqyRQmMfQA1AWIdfRMzC2tbB2dFFTV0D2JvP2h0AAtsCCtPIIBfVAVCGPVtXVRDY1NzSxt7J3zCuICvNh9-LP1SoxDK8JqogrcGz0T-VPTMwNKusOrIuoH4pugWuHHCIJQpqoja+X7YheHoACFVnKRc0EgYI+h6AA90enIAExh4PW5ZPgEhGgZmNgAJkkXBkvFo0QGxRQejKmx6kQhrn2jUOWUmFWm20cSPqB2aSVGGXRKFyKAheyKQVhnUxW16uPmqIJrUC7X0tNC9MRkJRQxZyTSxImwTpCKcjL5CQFsFOGK54pxvPczKWSROJNyuXwlHIwBwpmgABUNGAyLAdHowFY5rFihcxTMNLL0fCnRqJm7trL+IwdF6aidfToPYtAUgAw4XZ7HdtQzpo4sAMwR2M1RP+tMOeNIWUuerFE75qEhqySqlIABuAmgYDE5HoAHc4MVOd0ndb0dbbRXq8xaPWm8dirQbbsCzpRxSJ0gTbQtEA)

The results are below:

<table>
  <tr>
    <th>Test</th>
    <th>Java</th>
    <th>Typescript</th>
    <th>Dart</th>
  </tr>
  <tr>
    <td>Class variable is shared between parent and child class?</td>
    <td>Yes</td>
    <td>No (Shared till the first write to the variable using the child class)</td>
    <td> No </td>
  </tr>
  <tr>
    <td>Class variable can be accessed without the class name in a class method?</td>
    <td>Yes</td>
    <td>No (either use the class name or "this")</td>
    <td>Yes</td>
  </tr>
<tr>
<td>Class variable can be accessed without the class name in a object method?</td>
<td>Yes</td>
<td>No</td>
<td>Yes</td>
</tr>
<tr>
<td>Parent class variable can be accessed from a child class method without the class name?</td>
<td>Yes</td>
<td>No (either use the class name or "this")</td>
<td>No</td>
</tr>
<tr>
<td>Parent class variable can be accessed from a child object method without the class name?</td>
<td>Yes</td>
<td>No</td>
<td>No</td>
</tr>
<tr>
<td>Class variable accessible using an object?</td>
<td>Yes</td>
<td>No</td>
<td>No</td>
</tr>
<tr>
<td>Parent object method can get the value of a child class variable when called using a child object?</td>
<td>Yes</td>
<td>No</td>
<td>No</td>
</tr>
<tr>
<td>Child class variable hides the parent class variable?</td>
<td>Yes</td>
<td>Yes</td>
<td>Yes</td>
</tr>
<tr>
<td>Child class variable overrides the parent class variable?</td>
<td>No</td>
<td>No</td>
<td>No</td>
</tr>
</table>

I think we should follow the Dart behavior in Vim9 classes (class variables are not shared between a parent and a child class).

**Reply by errael** [reply]
I like it all except

> Child class variable hides the parent class variable?

Class member can be shadowed, but an object member can't? There are so many examples of where shadowing is illegal, why this one exception? Can the type be changed as well?

**Reply by yegappan** [reply]
In Dart, the type of a class variable in a child class can be different from the type of the same class variable in the parent class.

**Reply by errael** [reply]
> In Dart, the type of a class variable in a child class can be different

I understand that. But I'm asking about vim9script, not Dart. Is the plan to remove the restriction on shadowing everywhere? In other places?

**Reply by yegappan** [reply]
An object variable in a Vim9 child class should override the same object variable in the parent class.

**Reply by dkearns** [reply]
It's not the right question for this model as "parent" is meaningless.

```vim
vim9script
const ClassA = {
  foo: 11
}
# class relationship is _not_ relevant
const ClassB = {
  foo: "Twenty Two",
  another_foo: ClassA.foo
}
```

**Reply by errael** [reply]
So the statement

> Child class variable hides the parent class variable?

Is inaccurate. The class variable exists only as part of defining class, it has no relationship to child classes. Since it can't be accessed in the child with the bare name, there's nothing hidden. Thanks, I understand now.

**Reply by errael** [reply]
(@dkearns Is that Dart syntax)

**Reply by yegappan** [reply]
I have created the PR https://github.com/vim/vim/pull/13086 to implement these changes.  Please take a look and let me know if you see or run into any problems.

**Comment by errael** [comment]
About **static in an interface**

Consider that
1. a static cannot be accessed with an object 
2. a static can not be accessed with an interface name, "I.sval1"
3. a static in an interface must be defined in a class that implements it
4. a static is not inherited with class B extends A (a new decision, may have meaning here)

In Java, there's the _Constant Interface Antipattern_.
And because of 3. they can't even be used for sharing constants.
And Java insures that interface members are all constant.

I don't know what other languages do, but
this all suggests that a static in an interface should be illegal.

Haven't heard any preferences on this one way or the other.

**Reply by errael** [reply]
> static in an interface

@yegappan @dkearns comments? Does anyone think an interface (even one that's the result of specifies) should support static.

**Reply by yegappan** [reply]
I agree.  There is no need to support static variables in an interface.  In general, I am wondering should we support class/object public/private variables in an interface?  An interface will only support public class/object methods and constants.

**Reply by errael** [reply]
> support class/object public/private variables in an interface?
> An interface will only support public class/object methods and constants.

As discussed in #12987, particularly https://github.com/vim/vim/discussions/12987#discussioncomment-6880081, I believe a read-only object variable (not necessarily constant) is a way to specify a high performance property in an interface. Syntactically this agrees with no need for public/private members, but doesn't restrict the implementing class from modifying a read access member.

I favor
```
An interface will only support public class/object methods and read-only members.
```

I'm currently in the mood "don't put something in if you don't need it". Otherwise you'll have unused things being supported forever.

**Comment by errael** [comment]
About: class methods in an interface

I don't necessarily agree with supporting class methods in an interface, mostly ambivalent.

Consider
```
def F(i: Iface)
enddef
```
A class method can't be accessed through an object; it can't be accessed as Iface.method(). It can only be accessed with SomeClass.method(). The name of the class must be known, having it in an interface doesn't seem to have much use, except maybe as some kind of reminder. But I don't care much one way or the other.

**Reply by dkearns** [reply]
> A class method can't [...] be accessed as Iface.method(). It can only be accessed with SomeClass.method(). 

They should be accessed via the interface reference but this seems a moot point now as it appears that we're not getting static interface members.

**Comment by dkearns** [comment]
> (@dkearns Is that Dart syntax)

No, it's Vim9 dictionary syntax that I thought might be clearer than my ramblings.

**Comment by errael** [comment]
Closed as of patch 9.0.1906 https://github.com/vim/vim/commit/92d9ee5f4ca0d2de04c39afbafc7609da43fb2e9
